dll/win32/shell32/CDefaultContextMenu.cpp

```diff
@@ -513,53 +539,96 @@ CDefaultContextMenu::AddStaticContextMenusToMenu(
+        BOOL hide = FALSE;
         HKEY hkVerb;
         if (idResource > 0)
         {
-            if (LoadStringW(shell32_hInstance, idResource, wszVerb, _countof(wszVerb)))
-                mii.dwTypeData = wszVerb; /* use translated verb */
-            else
-                ERR("Failed to load string\n");
-
-            LONG res = RegOpenKeyW(info.hkClass, wszKey, &hkVerb);
-            if (res == ERROR_SUCCESS)
+            if (!(uFlags & CMF_OPTIMIZEFORINVOKE))
             {
-                res = RegQueryValueExW(hkVerb, L"Extended", NULL, NULL, NULL, NULL);
-                Extended = (res == ERROR_SUCCESS);
-
-                RegCloseKey(hkVerb);
+                if (LoadStringW(shell32_hInstance, idResource, wszVerb, _countof(wszVerb)))
+                    mii.dwTypeData = wszVerb; /* use translated verb */
+                else
+                    ERR("Failed to load string\n");
             }
+
+            if (RegOpenKeyW(info.hkClass, wszKey, &hkVerb) != ERROR_SUCCESS)
+                hkVerb = NULL;
         }
         else
         {
-            LONG res = RegOpenKeyW(info.hkClass, wszKey, &hkVerb);
-            if (res == ERROR_SUCCESS)
+            if (RegOpenKeyW(info.hkClass, wszKey, &hkVerb) == ERROR_SUCCESS)
             {
-                DWORD cbVerb = sizeof(wszVerb);
-                res = RegLoadMUIStringW(hkVerb, NULL, wszVerb, cbVerb, NULL, 0, NULL);
-                if (res == ERROR_SUCCESS)
+                if (!(uFlags & CMF_OPTIMIZEFORINVOKE))
                 {
-                    /* use description for the menu entry */
-                    mii.dwTypeData = wszVerb;
-                }
+                    DWORD cbVerb = sizeof(wszVerb);
 
-                res = RegQueryValueExW(hkVerb, L"Extended", NULL, NULL, NULL, NULL);
-                Extended = (res == ERROR_SUCCESS);
+                    LONG res = RegLoadMUIStringW(hkVerb, L"MUIVerb", wszVerb, cbVerb, NULL, 0, NULL);
+                    if (res || !*wszVerb)
+                        res = RegLoadMUIStringW(hkVerb, NULL, wszVerb, cbVerb, NULL, 0, NULL);
 
-                RegCloseKey(hkVerb);
+                    if (res == ERROR_SUCCESS && *wszVerb)
+                    {
+                        /* use description for the menu entry */
+                        mii.dwTypeData = wszVerb;
+                    }
+                }
+            }
+            else
+            {
+                hkVerb = NULL;
             }
         }
 
-        if (!Extended || GetAsyncKeyState(VK_SHIFT) < 0)
+        if (hkVerb)
         {
-            mii.cch = wcslen(mii.dwTypeData);
+            // FIXME: GetAsyncKeyState should not be called here, clients
+            // need to be updated to set the CMF_EXTENDEDVERBS flag.
+            if (!(uFlags & CMF_EXTENDEDVERBS) && GetAsyncKeyState(VK_SHIFT) >= 0)
+                hide |= RegValueExists(hkVerb, L"Extended");
+
+            hide |= RegValueExists(hkVerb, L"ProgrammaticAccessOnly");
```

Comment:
Just to be sure, the actual value stored inside these registry values do not actually matter? And what about their type (REG_SZ, REG_DWORD, etc.) ?

Reply:
Does not matter, they are boolean flags.

 > [The value should not have any data associated with it.](https://learn.microsoft.com/en-us/windows/win32/shell/context-menu-handlers#programmatic-access-only-verbs)

Reply:
OK but they still seem to insist on the `REG_SZ` part. Could you check on Windows whether the type is still checked?

Reply:
It does not matter, already checked.

Guess I should have quoted the docs for the other flag:

 > The value only needs to be present; it can be of any value type (though REG_NONE is most appropriate) and does not require any data to be associated with it. If data is present, it is ignored.